doc/example.clj

```diff
@@ -0,0 +1,44 @@
+(ns example
+  (:require
+    [bunnicula.component.connection :as rmq.connection]
+    [bunnicula.component.consumer-with-retry :as rmq.consumer]
+    [bunnicula.component.monitoring :as consumer.monitoring]
+    [bunnicula.component.publisher :as rmq.publisher]
+    [bunnicula.protocol :as rmq]
+    [clojure.tools.logging :as log]
+    [com.stuartsierra.component :as component]))
+
+
+;; connection details assume that the RabbitMQ is running, as setup in the
+;; docker-compose.yml file
+(defn create-system []
+  {:rmq-connection (rmq.connection/create {:host "127.0.0.1"
+                                           :port 5672
+                                           :username "rabbit"
+                                           :password "password"
+                                           :connection-name "example"
+                                           :vhost "/main"})
+   :monitoring (consumer.monitoring/create)
+   :publisher (component/using
+                (rmq.publisher/create)
+                [:rmq-connection])
+   :simple-consumer (component/using
+                      (rmq.consumer/create {:handler (fn [_ {:keys [number] :as _payload}
```

Comment:
shall we allow two fn arty as well for handler fns? We never endup using the other info in handlers.

Reply:
Don't think so, we'd have to somehow expose the function arity and [dynamically dispatch on that when invoking the handler function](https://github.com/nomnom-insights/nomnom.bunnicula/blob/95707fa38c00008f4da7b733a4ccf90192f5d116/src/bunnicula/component/consumer_with_retry.clj#L163). That might be slow or require a config option in the consumer component.

 If the handler was protocol/interface based, then it would be possible.

Reply:
Something to think about for V3, which I think could address this, and simplify usage with easy migration path:

- no monitoring component, but use middleware pattern like [Eternity](https://github.com/nomnom-insights/nomnom.eternity) for instrumentation (less setup)
- change handler signature to `[component body envelope]`, we can definitely drop the raw body (`bytes`) argument, it was never useful beyond migrating from JSON to Avro

```clojure

(def a-consumer 
(component/using 
  (consumer/create 
    {:options {:queue-name "some.other.queue.1" 
               :consumer-threads 3 
               :timeout-seconds 14 
               :retries 3 
               :prefetch 10}}
     :handler (-> bunnicula.middleware/error ;; would pull exception tracker out of deps
                  (bunnicula.middleware/log "a-consumer") ;; can return decorated closure with consumer name injected
                  bunnicula.middleware/stats ;; would use return value to create metrics
                  (fn [component body _envelope] ;; envelope is useful in middleware mostly, it could contain consumer name and the config too
                    (do-insert (:db-conn component) body)
                    :bunnicula.consumer/ack)))
  [:exception-tracker :statsd :rabbitmq/connection :db-conn])
```

The middlewares would work as expected, and use the consumer returns value to decide what to do etc etc etc. Much simpler approach 


- namespaced key for the connection component ( `:rabbitmq/connection` or something like that)
- remove non-namespaced consumer return flags (e.g. no more `:ack`, only `:bunnicula.consumer/ack` - more verbose, but also more explicit!)

- better serialization model for publishers: currently, we tied a publisher with a serializer, so if you need to push differently encoded Avro messages to different queues, you have to have multiple publishers. Instead we should keep one publisher component, but dispatch on the pushed queue name to invoke the serializer. Something like this:

```clojure

(defn create-publisher [config]
  (component/using
    (rmq.publisher/create 
      ;; optional binds serialization to queue name, uses JSON if no match
      {:serializers { "some.queue.number.1" (serializer.avro/build "schema/path/QueUenyumber1")
                      "some.other.queue" (serializer.avro/build "schema/path/SomeOtherQueue")
                      :else serializer.json/handler}})
    [:rabbitmq/connection]))


;; then

(rmq/publish publisher "some.other.queue" {:foo :bar}) ;; will serialize using configured serializer
(rmq/publishe publisher "blomp.flurps" {:timeline "sideways"}) ;; will use default serializer
```

That's off the top of my head, but having all of these will clean up a lot of code in both Bunnicula (no need for the whole monitoring component & protocol) and simplify application code too (fewer components in the system).

Reply:
Lastly, then you can have 2 arity function as the handler - you just need a middleware which strips out the last arg 😉